src/Connection.ts

```diff
@@ -149,45 +147,49 @@ class Connection extends EventEmitter {
    * @param {Protocol.Target.TargetInfo} targetInfo
    * @return {!Promise<!CDPSession>}
    */
-  async createSession(targetInfo) {
-    const {sessionId} = await this.send('Target.attachToTarget', {targetId: targetInfo.targetId, flatten: true});
+  async createSession(targetInfo: Protocol.Target.TargetInfo): Promise<CDPSession> {
+    const {sessionId} = await this.send<Protocol.Target.attachedToTargetPayload>('Target.attachToTarget', {targetId: targetInfo.targetId, flatten: true});
     return this._sessions.get(sessionId);
   }
 }
 
-class CDPSession extends EventEmitter {
-  /**
-   * @param {!Connection} connection
-   * @param {string} targetType
-   * @param {string} sessionId
-   */
-  constructor(connection, targetType, sessionId) {
+interface CDPSessionOnMessageObject {
+    id?: number;
+    method: string;
+    params: {};
+    error: {message: string; data: any};
+    result?: any;
+
+}
+export class CDPSession extends EventEmitter {
+  _connection: Connection;
+  _sessionId: string;
+  _targetType: string;
+  _callbacks: Map<number, ConnectionCallback> = new Map();
+
+  constructor(connection: Connection, targetType: string, sessionId: string) {
     super();
-    /** @type {!Map<number, {resolve: function, reject: function, error: !Error, method: string}>}*/
-    this._callbacks = new Map();
     this._connection = connection;
     this._targetType = targetType;
     this._sessionId = sessionId;
   }
 
-  /**
-   * @param {string} method
-   * @param {!Object=} params
-   * @return {!Promise<?Object>}
-   */
-  send(method, params = {}) {
+  send<T extends keyof Protocol.CommandParameters>(method: T, params?: Protocol.CommandParameters[T]): Promise<Protocol.CommandReturnValues[T]> {
     if (!this._connection)
       return Promise.reject(new Error(`Protocol error (${method}): Session closed. Most likely the ${this._targetType} has been closed.`));
-    const id = this._connection._rawSend({sessionId: this._sessionId, method, params});
+
+    const id = this._connection._rawSend({
+      sessionId: this._sessionId,
+      method,
+      params: params || {}
```

Comment:
hey @mjzffr I think there might be a Firefox bug here with the fact that it's expecting params to be present. Without this `|| {}` some Firefox specs fail like so:

```
 1) Page.click
       should not hang with touch-enabled viewports:
     Error: Protocol error (Page.reload): (destructured parameter) is undefined reload@chrome://remote/content/domains/content/Page.jsm:135:1
execute@chrome://remote/content/domains/DomainCache.jsm:99:25
receiveMessage@chrome://remote/content/sessions/ContentProcessSession.jsm:69:45
MessageListener.receiveMessage*ContentProcessSession@chrome://remote/content/sessions/ContentProcessSession.jsm:27:25
@chrome://remote/content/sessions/frame-script.js:12:1
      at Promise (lib/Connection.js:147:63)
      at new Promise (<anonymous>)
      at CDPSession.send (lib/Connection.js:146:16)
      at Page.reload (lib/Page.js:630:26)
      at Page.<anonymous> (lib/helper.js:82:27)
      at Page.setViewport (lib/Page.js:776:24)
      at process._tickCallback (internal/process/next_tick.js:68:7)
    -- ASYNC --
      at Page.<anonymous> (lib/helper.js:81:19)
      at Context.it (test/click.spec.js:207:16)
      at process._tickCallback (internal/process/next_tick.js:68:7)
```

However Chrome is fine. I think that maybe FF is expecting params where it shouldn't? This is an easy workaround for now but might be worth looking into. I also might have misunderstood the issue so sorry if that's the case. Also pinging @mathiasbynens for any thoughts.

Reply:
@jackfranklin Thanks, I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1631570